### ISSUE TITLE
internal error: redirect not found

### ISSUE THREAD

**Issue by tim-moody** [openzim/zimwriterfs#24]
running:

sudo docker build -t zimwriterfs zimwriterfs_config
time sudo docker run -it -v $(pwd)/articles:/articles -v $(pwd)/output:/output zimwriterfs

which runs

#!/bin/bash
zimwriterfs \
--title="CDC Extract" \
--description="Centers for Disease Control Web Site" \
--creator="www.cdc.gov" \
--publisher="Tim Moody" \
--tags="IIAB;Medical;CDC" \
--language="eng" \
--welcome="www.cdc.gov/index.html" \
--favicon="www.cdc.gov/images/home/favicon.ico" \
--withFullTextIndex \
--verbose \
/articles \
/output/cdc_en_novids_2017-10.zim \
 | tee /output/outp.log

on extract from www.cdc.gov 

processes for 70 minutes and
yields following messages and terminates

isArticleIndexer running: yes
isArticleIndexer running: no
sort 195094 directory entries (aid)
remove invalid redirects from 195094 directory entries
sort 195072 directory entries (url)
set index
translate redirect aid to index
internal error: redirect aid www.cdc.gov/globalhealth/healthprotection/ghsb/gphi/partnership.htm not found

but file exists
root@iiab-devel:~/zimwriterfs# ll articles/www.cdc.gov/globalhealth/healthprotection/ghsb/gphi/partnership.htm
-rw-r--r-- 1 root root 1606 Oct 24 14:19 articles/www.cdc.gov/globalhealth/healthprotection/ghsb/gphi/partnership.htm

**Comment by kelson42** [comment]
@tim-moody Would that be possible to make the content you want to zim somewhere available?

**Comment by tim-moody** [comment]
will do

**Comment by kelson42** [comment]
File with the content can be found at http://tmp.kiwix.org/24.zip

**Comment by kelson42** [comment]
@tim-moody There is not directory called "articles" in the ZIP file you have provided to us

**Comment by tim-moody** [comment]
Yes, 24.zip holds the directory www.cdc.gov and goes inside of articles.

The docker invocation above expects an articles directory and an output directory, both of which are mapped to directories of the same name inside the docker container.

**Comment by kelson42** [comment]
@tim-moody So, it tested myself, and we might agree that the error message or handling is not the best one but as far as I can see, for the rest, the hndlign is correct. The problem is that the data are wrong/corrupted.

Here is how ```articles/www.cdc.gov/globalhealth/healthprotection/ghsb/gphi/partnership.htm``` looks like:

```
<html><head><meta content="0; URL=https://www.cdc.gov/globalhealth/healthprotection/ghs/'http://www.cdc.gov/globalhealth/healthprotection/ghs/index.html'" http-equiv="refresh"/>
</head></html>
kelson@camber:/tmp/24$ emacs -nw articles/www.cdc.gov/globalhealth/healthprotection/ghs/index.htmkelson@camber:/tmp/24$ cat  articles/www.cdc.gov/globalhealth/healthprotection/ghsb/gphi/partnership.htm
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html id="no-js" lang="en-us" xml:lang="en-us" xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta content="Percussion CM System" name="generator"/>
<meta content="FmdgYGdlYHgHZGZ4AWNhYngGZmVgeBNnYGBkYG0=" id="perc_linkback" name="perc_linkback"/><!-- Internal Release Version # 3.6 -->
<meta content="IE=edge" http-equiv="X-UA-Compatible"/>
<meta content="text/html; charset=utf-8" http-equiv="content-type"/>
<meta content="CDC works 24/7 to save lives, reduce disease, and improve health around the world." name="description"/><meta content="CDC global health, Global health, center for global health, International health, International Health Regulations, International, Overseas, Global AIDS, Malaria, Influenza, Polio, Measles, FETP, FELTP, Safe Water, Emerging infections, Global disease detection, WHO, UNICEF, Sustainable management development, Public health capacity, Health systems strengthening, Global laboratory, Refugee health, Global travel, Kenya, Thailand,Guatemala, China, Egypt,  India, South Africa" name="keywords"/>
<meta content="0; URL=../../ghs/index.htm" http-equiv="refresh"/>
<meta content="en-us" name="DC.Language" scheme="ISO 639-1"/>
<meta content="2.7.0" name="template.version"/>
<meta content="255205" name="cdc:content-id"/>
<meta content="13" name="cdc:revision-id"/>
<link href="../../../../globalhealth/healthprotection/ghsb/index.htm" rel="canonical"/>
<title>CDC Global Health - Global Health Security</title>
</head>
<body>
</body>
</html>
```

Here is how  ```articles/www.cdc.gov/globalhealth/healthprotection/ghs/index.htm``` looks like:
```
index.htm
<html><head><meta content="0; URL=https://www.cdc.gov/globalhealth/healthprotection/ghs/'http://www.cdc.gov/globalhealth/healthprotection/ghs/index.html'" http-equiv="refresh"/>
</head></html>
```

... So the first redirects (in HTML) to the second and that looks quite coode... but the second is a redirect which is completly broken... and zimwriterfs can not deal with it and dies.

**Comment by tim-moody** [comment]
so is it the case that zimwriterfs can handle a broken link, but not a broken redirect? will any link to www.cdc.gov or other external site kill it?

or is it the syntax of   

URL=https://www.cdc.gov/globalhealth/healthprotection/ghs/'http://www.cdc.gov/globalhealth/healthprotection/ghs/index.html'

**Comment by kelson42** [comment]
@tim-moody I do not know, but this has to be tested in detail.

**Comment by tim-moody** [comment]
OK. Thanks. I will normalize the redirects and rerun. FWIW the chrome browser actually handles this redirect correctly inspite of the messed up syntax.

**Comment by tim-moody** [comment]
**Test 1:**

in articles/www.cdc.gov/globalhealth/healthprotection/ghs/index.htm change
`<meta content="0; URL=https://www.cdc.gov/globalhealth/healthprotection/ghs/'http://www.cdc.gov/globalhealth/healthprotection/ghs/index.html'" http-equiv="refresh"/>` to
`<meta content="0; URL=https://www.cdc.gov/globalhealth/healthprotection/ghs/index.html" http-equiv="refresh"/>`

This yields the same error message and halt as first reported.

**Comment by tim-moody** [comment]
**Test 2:**

in articles/www.cdc.gov/globalhealth/healthprotection/ghs/index.htm change to
`<meta content="0; URL=index.html" http-equiv="refresh"/>`

Now zimwriterfs completes with the following messages:

isArticleIndexer running: yes
isArticleIndexer running: no
sort 195095 directory entries (aid)
remove invalid redirects from 195095 directory entries
sort 195074 directory entries (url)
set index
translate redirect aid to index
195074 directory entries created
create title index
195074 title index created
4440 clusters created
fill header
write zimfile
ready

real    72m54.561s
user    0m0.716s
sys     0m1.092s

**Comment by tim-moody** [comment]
This puzzles me a bit in that there are approximately 140 other instances of external redirects that did not cause a problem. So it seems:

- external redirect - no problem
- internal redirect to page with external redirect - zimwriterfs stops
- internal redirect to page with internal redirect - no problem

**Comment by kelson42** [comment]
Yes, quite clear something here is fishy... that said, have redirect to external link is a "non-sense" for an offline content.

**Comment by asashnov** [comment]
It may be non-actual for now, as the page www.cdc.gov/globalhealth/healthprotection/ghsb/gphi/partnership.htm doesn't contain this broken syntax now:

```
<meta http-equiv="refresh" content="0; url=/globalhealth/healthprotection/ghs/index.htm">
```

**Comment by kelson42** [comment]
@asashnov Why you look online? This is not relevant. The only relevant thing is what is on the directory to package in a ZIM file (and sorry but http://tmp.kiwix.org/24.zip looks like has been deleted, but you have the important details in the comments).

**Comment by mgautierfr** [comment]
I don't know what we have to fix here.
It seems normal to me that zimwriterfs fails to create a redirect to an external content.

**Comment by kelson42** [comment]
Yes,butwe need a prper error handling.

**Comment by mgautierfr** [comment]
The redirection error message has changed since the issue has been opened (https://github.com/openzim/libzim/commit/7a67b5145ba08832525684e65fb12923fc5726eb#diff-77ad09569772228e61411d708e72343fR594).
